File: Integer/SumArrayK/SumArrayK.cc
class SumArrayK{
	public:
		vector<int> sumArrayK(int[] a, int k) {
			vector<int> result;
			
			map<int, int> hash;
			
			hash[0] = 0;
			int curr_sum = 0;
			
			for(int i = 0; i < a.size(); ++i) {
				curr_sum += a[i];
				if(hash.find(curr_sum - k) != hash.end()) {
					result.push_back(hash(curr_sum - k));
					result.push_back(i);
					return result;					
				} else {
					hash[curr_sum] = i + 1;
				}
				
			}
			return result;
		}
		
		/*Time: O(2n), Space: O(1)*/
		vector<int> sumArrayK1(int[] a, int k) {
			vector<int> result;
			
			int left_index = 0, curr_sum = 0;
			for(int i = 0; i < a.size(); ++i) {
				while(curr_sum > k) {
					curr_sum -= a[left_index];
					++left_index;
				}
				
				if(curr_sum == k) {
					result.push_back(a[left_index]);
					result.push_back(i);
					return result;
				}
				curr_sum += a[i];
			}
			
			return result;
		}
};

File: Integer/SumArrayK/SumArrayK.java
class SumArrayK {
	
	/*O(n^2)*/
	public boolean sumArrayK(int[] a, int n, int sum) {
		int i, j, curr_sum;
		
		for(i = 0; i < n; ++i) {
			curr_sum = a[i];
			for(j = i; j < n; ++j) {
				if(curr_sum == sum) {
					return true;
				}
				if(curr_sum > sum || j == n) {
					break;
				}
				curr_sum += a[i];
			}
		}
		return false;
	}
	
	/*O(n)*/
	public boolean sumArrayK1(int[] , int n, int sum) {
		int start = 0, j, curr_sum = a[0];
	
		for(int i = 0; i < n; ++i) {
			while(curr_sum > sum && start < i - 1) {
				curr_sum -= a[start];
				start++;
			}
			
			if(curr_sum == sum)
				return true;
			
			if(i < n)
				curr_sum += a[i];
			
			
			if(curr_sum > sum || i == n)
				return false;
			
			
		}
		return false;
	}
}